src/coreclr/jit/morph.cpp

```diff
@@ -12377,6 +12377,29 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
     GenTree* op1 = add->gtGetOp1();
     GenTree* op2 = add->gtGetOp2();
 
+    // Fold "GTF_ICON_PINNED_HDL + cns"
+    if (op1->IsCnsIntOrI() && op2->IsCnsIntOrI())
+    {
+        GenTreeIntCon* op1Cns = op1->AsIntCon();
+        GenTreeIntCon* op2Cns = op2->AsIntCon();
+
+        if (op1->IsIconHandle(GTF_ICON_PINNED_HDL))
```

Comment:
I do not think adding a constant to the icon and calling it a new icon is going to work well. Icons are like handles. 

The right place to deal with this is emit - the offset should be passed in as delta for `recordRelocation`.

Reply:
I am not sure I understand - there is no relocation needed in my snippet above ^ where I hit `ADD(constStrPinnedHandle, 12)` pattern to fold
But I've just realized I indeed don't need a separate flag here and can do folding for any `GTF_ICON_STR_HDL` (but will have to be careful with debug display - pointers might not point to the start of a string)

Reply:
The JITing at runtime does not typically need to map Icon backs to what they were. There are exception like the JIT dump problem you run into.

AOT mode needs to map icon backs to what they were a lot more often.

Reply:
Ah, makes sense, removed that logic, will check separately if it can be safely enabled for JIT
For tdoay I am only interested to make it green - locally it passes my tests 🙂

Reply:
For the JIT case you may want to use the same mapping as `GetFoldedArithOpResultHandleFlags` does in VN. It recasts the handle as more general "no-knowledge" handle types.

Reply:
> For the JIT case you may want to use the same mapping as `GetFoldedArithOpResultHandleFlags` does in VN. It recasts the handle as more general "no-knowledge" handle types.

yeah, this one here also helped with unexpected re-ordering "CONST_HDL op CNS" before VN - caused some problems for gtMarkAddrMode 🙂